### PR TITLE
Keep `docker` Docker images updated (#5298)

### LINKED ISSUE #5298: Keep `docker` Docker images updated
Every six months or when critical or high CVE findings make it necessary, the operator will file a PR that updates Azul to the latest version of `docker.io/library/docker:${version}` and `docker.io/library/docker:${version}-dind`.

### PR DESCRIPTION
**Pull request by dsotirho-ucsc**
Connected issues: #5298

### Author

- [x] PR is a draft
- [x] Target branch is `develop`
- [x] Name of PR branch matches `issues/<GitHub handle of author>/<issue#>-<slug>`
- [x] PR title references all connected issues
- [x] PR title matches<sup>1</sup> that of a connected issue <sub>or comment in PR explains why they're different</sub>
- [x] For each connected issue, there is at least one commit whose title references that issue
- [x] PR is connected to all connected issues via ZenHub
- [x] PR description links to connected issues
- [x] Added `partial` label to PR <sub>or this PR completely resolves all connected issues</sub>

<sup>1</sup> when the issue title describes a problem, the corresponding PR
title is `Fix: ` followed by the issue title

### Author (reindex, API changes)

- [x] Added `r` tag to commit title <sub>or this PR does not require reindexing</sub>
- [x] Added `reindex` label to PR <sub>or this PR does not require reindexing</sub>
- [x] Added `a` (compatible changes) or `A` (incompatible ones) tag to commit title <sub>or this PR does not modify the Azul service API</sub>
- [x] Added `API` label to connected issues <sub>or this PR does not modify the Azul service API</sub>

### Author (chains)

- [x] This PR is blocked by previous PR in the chain <sub>or this PR is not chained to another PR</sub>
- [x] Added `base` label to the blocking PR <sub>or this PR is not chained to another PR</sub>
- [x] Added `chained` label to this PR <sub>or this PR is not chained to another PR</sub>

### Author (upgrading)

- [x] Documented upgrading of deployments in UPGRADING.rst <sub>or this PR does not require upgrading</sub>
- [x] Added `u` tag to commit title <sub>or this PR does not require upgrading</sub>
- [x] Added `upgrade` label to PR <sub>or this PR does not require upgrading</sub>

### Author (operator tasks)

- [x] Added checklist items for additional operator tasks <sub>or this PR does not require additional tasks</sub>

### Author (hotfixes)

- [x] Added `F` tag to main commit title <sub>or this PR does not include permanent fix for a temporary hotfix</sub>
- [x] Reverted the temporary hotfixes for any connected issues <sub>or the `prod` branch has no temporary hotfixes for any connected issues</sub>

### Author (before every review)

- [x] Rebased PR branch on `develop`, squashed old fixups
- [x] Ran `make requirements_update` <sub>or this PR does not touch requirements*.txt, common.mk, Makefile and Dockerfile</sub>
- [x] Added `R` tag to commit title <sub>or this PR does not touch requirements*.txt</sub>
- [x] Added `reqs` label to PR <sub>or this PR does not touch requirements*.txt</sub>
- [x] `make integration_test` passes in personal deployment <sub>or this PR does not touch functionality that could break the IT</sub>

### Peer reviewer (after requesting changes)

Uncheck the *Author (before every review)* checklists.

### Peer reviewer (after approval)

- [x] PR is not a draft
- [x] Ticket is in *Review requested* column
- [x] Requested review from primary reviewer
- [x] Assigned PR to primary reviewer

### Primary reviewer (after requesting changes)

Uncheck the *before every review* checklists. Update the `N reviews` label.

### Primary reviewer (after approval)

- [x] Actually approved the PR
- [x] Labeled connected issues as `demo` or `no demo`
- [x] Commented on connected issues about demo expectations <sub>or all connected issues are labeled `no demo`</sub>
- [x] Decided if PR can be labeled `no sandbox`
- [x] PR title is appropriate as title of merge commit
- [x] `N reviews` label is accurate
- [x] Moved ticket to *Approved* column
- [x] Assigned PR to current operator

### Operator (before pushing merge the commit)

- [x] Checked `reindex` label and `r` commit title tag
- [x] Checked that demo expectations are clear <sub>or all connected issues are labeled `no demo`</sub>
- [x] PR has checklist items for upgrading instructions <sub>or PR is not labeled `upgrade`</sub>
- [x] Squashed PR branch and rebased onto `develop`
- [x] Sanity-checked history
- [x] Pushed PR branch to GitHub
- [x] Pushed PR branch to GitLab `dev` and added `sandbox` label <sub>or PR is labeled `no sandbox`</sub>
- [x] Pushed PR branch to GitLab `anvildev` <sub>or PR is labeled `no sandbox`</sub>
- [x] Pushed PR branch to GitLab `anvilprod` <sub>or PR is labeled `no sandbox`</sub>
- [x] Build passes in `sandbox` deployment <sub>or PR is labeled `no sandbox`</sub>
- [x] Build passes in `anvilbox` deployment <sub>or PR is labeled `no sandbox`</sub>
- [x] Build passes in `hammerbox` deployment <sub>or PR is labeled `no sandbox`</sub>
- [x] Reviewed build logs for anomalies in `sandbox` deployment <sub>or PR is labeled `no sandbox`</sub>
- [x] Reviewed build logs for anomalies in `anvilbox` deployment <sub>or PR is labeled `no sandbox`</sub>
- [x] Reviewed build logs for anomalies in `hammerbox` deployment <sub>or PR is labeled `no sandbox`</sub>
- [x] Deleted unreferenced indices in `sandbox` <sub>or this PR does not remove catalogs or otherwise causes unreferenced indices </sub>
- [x] Deleted unreferenced indices in `anvilbox` <sub>or this PR does not remove catalogs or otherwise causes unreferenced indices </sub>
- [x] Deleted unreferenced indices in `hammerbox` <sub>or this PR does not remove catalogs or otherwise causes unreferenced indices </sub>
- [x] Started reindex in `sandbox` <sub>or this PR does not require reindexing `sandbox`</sub>
- [x] Started reindex in `anvilbox` <sub>or this PR does not require reindexing `sandbox`</sub>
- [x] Started reindex in `hammerbox` <sub>or this PR does not require reindexing `sandbox`</sub>
- [x] Checked for failures in `sandbox` <sub>or this PR does not require reindexing `sandbox`</sub>
- [x] Checked for failures in `anvilbox` <sub>or this PR does not require reindexing `sandbox`</sub>
- [x] Checked for failures in `hammerbox` <sub>or this PR does not require reindexing `sandbox`</sub>
- [x] Title of merge commit starts with title from this PR
- [x] Added PR reference to merge commit title
- [x] Added commit title tags to merge commit title
- [x] Moved connected issues to Merged column in ZenHub
- [x] Pushed merge commit to GitHub

### Operator (chain shortening)

- [x] Changed the target branch of the blocked PR to `develop` <sub>or this PR is not labeled `base`</sub>
- [x] Removed the `chained` label from the blocked PR <sub>or this PR is not labeled `base`</sub>
- [x] Removed the blocking relationship from the blocked PR <sub>or this PR is not labeled `base`</sub>
- [x] Removed the `base` label from this PR <sub>or this PR is not labeled `base`</sub>

### Operator (after pushing the merge commit)

- [x] Deployed `dev.shared`
- [x] Deployed `dev.gitlab`
- [x] Ran `make -C terraform/gitlab/runner` to deploy GitLab runner to `dev`.
- [x] Pushed merge commit to GitLab `dev` <sub>or PR is labeled `no sandbox`</sub>
- [x] Deployed `anvildev.shared`
- [x] Deployed `anvildev.gitlab`
- [x] Ran `make -C terraform/gitlab/runner` to deploy GitLab runner to `anvildev`.
- [x] Pushed merge commit to GitLab `anvildev` <sub>or PR is labeled `no sandbox`</sub>
- [x] Deployed `anvilprod.shared`
- [x] Deployed `anvilprod.gitlab`
- [x] Ran `make -C terraform/gitlab/runner` to deploy GitLab runner to `anvilprod`.
- [x] Pushed merge commit to GitLab `anvilprod` <sub>or PR is labeled `no sandbox`</sub>
- [x] Build passes on GitLab `dev`<sup>1</sup>
- [x] Reviewed build logs for anomalies on GitLab `dev`<sup>1</sup>
- [x] Build passes on GitLab `anvildev`<sup>1</sup>
- [x] Reviewed build logs for anomalies on GitLab `anvildev`<sup>1</sup>
- [x] Build passes on GitLab `anvilprod`<sup>1</sup>
- [x] Reviewed build logs for anomalies on GitLab `anvilprod`<sup>1</sup>
- [x] Deleted PR branch from GitHub
- [x] Deleted PR branch from GitLab `dev`
- [x] Deleted PR branch from GitLab `anvildev`
- [x] Deleted PR branch from GitLab `anvilprod`

<sup>1</sup> When pushing the merge commit is skipped due to the PR being
labelled `no sandbox`, the next build triggered by a PR whose merge commit *is*
pushed determines this checklist item.

### Operator (reindex)

- [x] Deleted unreferenced indices in `dev` <sub>or this PR does not remove catalogs or otherwise causes unreferenced indices </sub>
- [x] Deleted unreferenced indices in `anvildev` <sub>or this PR does not remove catalogs or otherwise causes unreferenced indices </sub>
- [x] Deleted unreferenced indices in `anvilprod` <sub>or this PR does not remove catalogs or otherwise causes unreferenced indices </sub>
- [x] Started reindex in `dev` <sub>or this PR does not require reindexing</sub>
- [x] Started reindex in `anvildev` <sub>or this PR does not require reindexing</sub>
- [x] Started reindex in `anvilprod` <sub>or this PR does not require reindexing</sub>
- [x] Checked for and triaged indexing failures in `dev` <sub>or this PR does not require reindexing</sub>
- [x] Checked for and triaged indexing failures in `anvildev` <sub>or this PR does not require reindexing</sub>
- [x] Checked for and triaged indexing failures in `anvilprod` <sub>or this PR does not require reindexing</sub>
- [x] Emptied fail queues in `dev` deployment <sub>or this PR does not require reindexing</sub>
- [x] Emptied fail queues in `anvildev` deployment <sub>or this PR does not require reindexing</sub>
- [x] Emptied fail queues in `anvilprod` deployment <sub>or this PR does not require reindexing</sub>

### Operator

- [x] Added checklist items to `prod` promotion PR:
> - [ ] Deployed `prod.shared`
> - [ ] Deployed `prod.gitlab`
> - [ ] Ran `make -C terraform/gitlab/runner` to deploy GitLab runner to `prod`.

- [x] Unassigned PR

## Shorthand for review comments

- `L` line is too long
- `W` line wrapping is wrong
- `Q` bad quotes
- `F` other formatting problem

### CODE REVIEW

**Comment by dsotirho-ucsc** [comment]
Tested with `make -C terraform/gitlab/runner docker` to build image locally.

**Review by achave11-ucsc — CHANGES_REQUESTED** [review]
I think this should be further tested, maybe deploying and changing the infra in GitLab `dev`, and triggering a `sandbox` build to ensure everything works good after the changes. If something goes terrible wrong (it'll be good if we caught it early before merging the PR branch), and  you could recover by recreating a volume from a recent snapshot. Building the image doesn't yield information about the new docker distro and its potential non-backward compatible changes.

**Comment by dsotirho-ucsc** [comment]
> I think this should be further tested, maybe deploying and changing the infra in GitLab `dev`, and triggering a `sandbox` build to ensure everything works good after the changes

Deployed `shared`, `gitab`, and `gitlab/runner` to `dev` from the PR branch, and updated the installed packages (yum update) on the instance. After a reboot, the jobs in the latest [pipeline](https://gitlab.dev.singlecell.gi.ucsc.edu/ucsc/azul/-/pipelines/25154) on GitLab `dev` were re-run. 

No problems encountered or found in the build logs.

**Review by achave11-ucsc — DISMISSED** [review]
Nicely done.

File: UPGRADING.rst [hunk 1]
```diff
@@ -20,6 +20,18 @@ reverted. This is all fairly informal and loosely defined. Hopefully we won't
 have too many entries in this file.
 
 
+#5298 Keep docker Docker images updated
+=======================================
+
+Operator
+~~~~~~~~
+
+Manually deploy the ``shared`` & ``gitlab`` components of any main deployment,
```
Comment (hannes-ucsc) [inline UPGRADING.rst]:
```suggestion
Manually deploy the ``shared`` & ``gitlab`` components (in that order) of any main deployment,
```

File: terraform/gitlab/runner/Dockerfile [hunk 2]
```diff
@@ -1,5 +1,5 @@
 ARG registry
-FROM ${registry}docker.io/library/docker:20.10.18
+FROM ${registry}docker.io/library/docker:24.0.2
 
 RUN apk -U add make bash docker-credential-ecr-login
```
Comment (hannes-ucsc) [inline terraform/gitlab/runner/Dockerfile]: I think we need `apk update` and `upgrade` here as well. And the ARG that forces the layer, as in the other PR.

File: terraform/gitlab/gitlab.tf.json.template.py [hunk 3]
```diff
@@ -324,7 +324,7 @@ def remove_inconsequential_statements(statements: list[JSON]) -> list[JSON]:
 # deploying the `gitlab` component.
 
 clamav_image = config.docker_registry + 'docker.io/clamav/clamav:1.1.0-1'
-dind_image = config.docker_registry + 'docker.io/library/docker:20.10.18-dind'
+dind_image = config.docker_registry + 'docker.io/library/docker:24.0.2-dind'
```
Comment (hannes-ucsc) [inline terraform/gitlab/gitlab.tf.json.template.py]: And you are sure that you observed this version actually running on the instance during the experiment that @achave11-ucsc suggested? You did redeploy the `gitlab` component during that experiment, and it was replaced, correct?

**Review by hannes-ucsc — CHANGES_REQUESTED** [review]

File: terraform/gitlab/gitlab.tf.json.template.py (hunk 3 above)
Comment (dsotirho-ucsc) [reply terraform/gitlab/gitlab.tf.json.template.py]: Yes. The deploy of gitlab and gitlab/runner was reviewed from my terminal history during PL.

**Comment by dsotirho-ucsc** [comment]
Tested on `dev`. Evidence of `apk update / upgrade` was observed during the deploy of gitlab/runner.

<details>
<summary>Expand for terminal log</summary>

```
(.venv) daniel@Crispin ~/repo/azul1 $ CI_COMMIT_REF_NAME=develop make -C terraform/gitlab/runner
docker build \
	       --platform linux/amd64 \
	       --build-arg registry=122796619775.dkr.ecr.us-east-1.amazonaws.com/ \
	       --tag docker.gitlab.dev.singlecell.gi.ucsc.edu/ucsc/azul/runner .
[+] Building 6.6s (9/9) FINISHED                                                                                                                  docker:desktop-linux
 => [internal] load .dockerignore                                                                                                                                 0.0s
 => => transferring context: 2B                                                                                                                                   0.0s
 => [internal] load build definition from Dockerfile                                                                                                              0.0s
 => => transferring dockerfile: 2.08kB                                                                                                                            0.0s
 => [internal] load metadata for 122796619775.dkr.ecr.us-east-1.amazonaws.com/docker.io/library/docker:24.0.2                                                     0.9s
 => [auth] sharing credentials for 122796619775.dkr.ecr.us-east-1.amazonaws.com                                                                                   0.0s
 => CACHED [1/4] FROM 122796619775.dkr.ecr.us-east-1.amazonaws.com/docker.io/library/docker:24.0.2@sha256:5c705f988236dc82bdc3b40f0ce57f211cb33c3e5a69c2b979a349  0.0s
 => [2/4] RUN apk update     && apk upgrade                                                                                                                       3.2s
 => [3/4] RUN apk -U add make bash docker-credential-ecr-login                                                                                                    2.0s
 => [4/4] RUN mkdir -p ${HOME}/.docker     && printf '{"credHelpers": {"%s": "ecr-login"}}\n' "${registry%/}"     > "${HOME}/.docker/config.json"                 0.3s
 => exporting to image                                                                                                                                            0.1s
 => => exporting layers                                                                                                                                           0.1s
 => => writing image sha256:94fcad7d01c15d241ba7326da7aa49c6166e988481ce2a1757bc549f189f0b9d                                                                      0.0s
 => => naming to docker.gitlab.dev.singlecell.gi.ucsc.edu/ucsc/azul/runner                                                                                        0.0s

What's Next?
  View summary of image vulnerabilities and recommendations → docker scout quickview
docker push docker.gitlab.dev.singlecell.gi.ucsc.edu/ucsc/azul/runner
Using default tag: latest
The push refers to repository [docker.gitlab.dev.singlecell.gi.ucsc.edu/ucsc/azul/runner]
ec1ac3a96e7f: Pushed
7fb114632c46: Pushed
cb55d9ddadee: Pushed
25cfa3861c90: Layer already exists
b2171bef574b: Layer already exists
c7ba5fc6cf01: Layer already exists
8414679fe4da: Layer already exists
58b5cf42218c: Layer already exists
2c97ae14af68: Layer already exists
38d4031c10e4: Layer already exists
d8935a465ee5: Layer already exists
9741b1546733: Layer already exists
29512f91e94b: Layer already exists
00db86641510: Layer already exists
aa30a02584e8: Layer already exists
d8ae153f9f36: Layer already exists
78a822fe2a2d: Layer already exists
latest: digest: sha256:cad73e4554f898ab5b8c1bdd8aaddc92eab0255fb58f1e69893952fdc30955f8 size: 3880
```

</details>

**Review by hannes-ucsc — APPROVED** [review]

File: terraform/gitlab/runner/Dockerfile [hunk 4]
```diff
@@ -1,5 +1,11 @@
 ARG registry
-FROM ${registry}docker.io/library/docker:20.10.18
+FROM ${registry}docker.io/library/docker:24.0.2
+
+# Increment this to update the OS packages
```
Comment (hannes-ucsc) [inline terraform/gitlab/runner/Dockerfile]:
```suggestion
# Increment the value of this variable to ensure that all installed OS packages are upgraded.
```

with wrapping.

File: terraform/gitlab/runner/Dockerfile [hunk 5]
```diff
@@ -1,5 +1,11 @@
...
+#
+ENV azul_docker_executor_image_version=1
+RUN apk update \
+    && apk upgrade
 
 RUN apk -U add make bash docker-credential-ecr-login
```
Comment (hannes-ucsc) [inline terraform/gitlab/runner/Dockerfile]:
```suggestion
ENV azul_docker_executor_image_version=1
RUN apk -U upgrade \
    && apk add make bash docker-credential-ecr-login
```

**Review by hannes-ucsc — CHANGES_REQUESTED** [review]

**Review by hannes-ucsc — APPROVED** [review]

**Review by hannes-ucsc — CHANGES_REQUESTED** [review]
Conflicts.

**Review by hannes-ucsc — APPROVED** [review]

**Pull request merged**